File: packages/sdks/server-sdks/gatrix-dotnet-server-sdk/src/Gatrix.Server.Sdk/Evaluation/MurmurHash3.cs
using System.Text;
using Murmur;

namespace Gatrix.Server.Sdk.Evaluation;

/// <summary>
/// MurmurHash3 wrapper using the murmurhash NuGet package (darrenkopp).
/// Used for sticky rollout percentage calculation — must produce
/// identical results to the npm "murmurhash" v3 package used in @gatrix/shared.
/// </summary>
public static class MurmurHash3
{
    /// <summary>
    /// Compute MurmurHash3 32-bit hash of a string key with a given seed.
    /// Matches the npm murmurhash v3 package output.
    /// </summary>
    public static uint Hash(string key, uint seed = 0)
    {
        return Hash(key.AsSpan(), seed);
    }

    /// <summary>
    /// Compute MurmurHash3 32-bit hash of a ReadOnlySpan<char> with a given seed without allocating strings.
    /// </summary>
    public static uint Hash(ReadOnlySpan<char> key, uint seed = 0)
    {
        int maxByteCount = Encoding.UTF8.GetMaxByteCount(key.Length);
        byte[]? rented = null;
        Span<byte> utf8Bytes = maxByteCount <= 512 
            ? stackalloc byte[maxByteCount] 
            : (rented = System.Buffers.ArrayPool<byte>.Shared.Rent(maxByteCount));

        try
        {
            int bytesWritten = Encoding.UTF8.GetBytes(key, utf8Bytes);
            using var algorithm = MurmurHash.Create32(seed);

            Span<byte> hashBytes = stackalloc byte[4];
            if (algorithm.TryComputeHash(utf8Bytes[..bytesWritten], hashBytes, out _))
            {
                return BitConverter.ToUInt32(hashBytes);
            }
            // Fallback just in case
            return BitConverter.ToUInt32(algorithm.ComputeHash(utf8Bytes[..bytesWritten].ToArray()));
        }
        finally
        {
            if (rented != null)
                System.Buffers.ArrayPool<byte>.Shared.Return(rented);
        }
    }
}

File: packages/sdks/server-sdks/gatrix-java-server-sdk/src/main/kotlin/com/gatrix/server/sdk/evaluation/MurmurHash3.kt
package com.gatrix.server.sdk.evaluation

import com.google.common.hash.Hashing
import java.nio.charset.StandardCharsets

/**
 * MurmurHash3 utility wrapper around Guava's Hashing.murmur3_32_fixed().
 * Must produce identical results to npm "murmurhash" v3 package used in @gatrix/evaluator.
 */
object MurmurHash3 {

    /**
     * Compute MurmurHash3 32-bit hash of a string with a given seed.
     * Returns unsigned 32-bit value as Long.
     */
    fun hash(key: String, seed: Int = 0): Long {
        val hashFunction = Hashing.murmur3_32_fixed(seed)
        val hashCode = hashFunction.hashString(key, StandardCharsets.UTF_8)
        // Convert to unsigned long
        return hashCode.asInt().toLong() and 0xFFFFFFFFL
    }

    /**
     * Compute a normalized hash (0-100) from stickiness key and group ID.
     * Used for rollout percentage calculation.
     */
    fun normalizedHash(stickinessKey: String, groupId: String, seed: Int = 0): Int {
        val combined = "$groupId:$stickinessKey"
        val hashValue = hash(combined, seed)
        return (hashValue % 101).toInt()
    }
}
